src/utils/helpers/auto-focus/auto-focus.js

```diff
@@ -0,0 +1,24 @@
+import guid from '../guid';
+
+const AutoFocus = {
+  default: (key, autoFocus) => {
+    if (key && autoFocus) {
+      return { key, autoFocus };
+    }
+
+    return {
+      key: guid(),
+      autoFocus: false
+    };
+  },
+
+  getKey: (autoFocus, { ...previous }) => {
+    if (autoFocus && previous && previous.autoFocus === autoFocus) {
+      return previous.key;
+    }
+
+    return guid();
```

Comment:
Since the second parameter of `AutoFocus.getKey()` isn't supplied in any call to this function in this PR, it looks like this function will always return a new unique ID from `guid()` every time it's called. (The first parameter `autoFocus` will have no effect if the second parameter isn't supplied.)

I'm not sure whether this is correct?

I know @jamime introduced similar functionality in #2540 to force the `Decimal` component to re-render when the `allowEmptyValue` Storybook knob is toggled.

Is the idea here to force the component to re-render when the `autoFocus` Storybook knob is toggled? (So that if `autoFocus` is now `true`, then the component will auto-focus when it's re-rendered?)

Reply:
Yes, toggle on knob should force re-render.

Reply:
OK. In that case, since the second parameter `{ ...previous }` is currently never supplied in any runtime/production code within the Carbon library, then this condition is currently always `false`:
```js
if (autoFocus && previous && previous.autoFocus === autoFocus) {
```
and so this `AutoFocus.getKey()` function is always behaving like this in runtime/production code:
```js
getKey: (autoFocus) => {
  return guid();
}
```
I'm guessing that the usage of `AutoFocus.getKey()` needs to be updated in some/all of these places:

* https://github.com/blaky/carbon/blob/0cb3002/src/__experimental__/components/checkbox/checkbox.stories.js#L91
* https://github.com/blaky/carbon/blob/0cb3002/src/__experimental__/components/select/select.stories.js#L39
* https://github.com/blaky/carbon/blob/0cb3002/src/__experimental__/components/textarea/textarea.stories.js#L62
* https://github.com/blaky/carbon/blob/0cb3002/src/__experimental__/components/textbox/textbox.stories.js#L150

to pass a value for the second parameter, so that this `AutoFocus.getKey()` function can return `previous.key` when appropriate?

Reply:
updated stories for those components.

Reply:
Thanks for adding values for the second parameter :white_check_mark: b19a82d